### PR TITLE
WL-2777 Made append path to work with pages and subsites.

### PR DESCRIPTION
**Pull request by ouit0408**
Added new class 'Redirect' in package portal which has redirect url and
appendPath attributes.
getRedirect() method in HierarchySiteNeighbourhoodService
also sets redirectUrl for subsites which has appendPath set as true.
It returns 'Redirect' object containing relevant info which will be used by
calling method in 'siteHandler' class to recreate redirect url with tool
information.

### CODE REVIEW

File: tetraelf-hierarchy/hierarchy-portal/impl/src/java/org/sakaiproject/portal/impl/HierarchySiteNeighbourhoodService.java [hunk 1]
```diff
@@ -177,17 +179,50 @@ public String parseSiteAlias(String alias) {
...
+						//save the trimmed path if there is a subsite redirect
+						trimmedPath = (StringUtils.isBlank(trimmedPath)) ? tempString : (tempString + trimmedPath) ;
+						siteId = firstPart;
+					}
+				}
+				else{
+					//node found skip the loop
+					break;
+				}
+			}while(StringUtils.countMatches(siteId, PortalHierarchyService.SEPARATOR) > 1);
```
Comment (buckett) [inline tetraelf-hierarchy/hierarchy-portal/impl/src/java/org/sakaiproject/portal/impl/HierarchySiteNeighbourhoodService.java]: This is a risky pattern as if there is a problem in the code we can get runaway threads, it's much safer to have a known loop.

File: tetraelf-hierarchy/hierarchy-portal/impl/src/java/org/sakaiproject/portal/impl/HierarchySiteNeighbourhoodService.java [hunk 2]
```diff
@@ -177,17 +179,50 @@ public String parseSiteAlias(String alias) {
...
+				if(appendPath && StringUtils.isNotBlank(trimmedPath)){
+					redirect.setUrl(redirectUrl + trimmedPath);
+				}
+				else{
+					redirect.setUrl(redirectUrl);
+				}
+				return redirect;
 			}
 			// If we have trailing seperators pull them off
 			Matcher matcher = trimTailingSeparators.matcher(siteId);
```
Comment (buckett) [inline tetraelf-hierarchy/hierarchy-portal/impl/src/java/org/sakaiproject/portal/impl/HierarchySiteNeighbourhoodService.java]: I think we probably want todo this check first as it's simpler and if we have trailing separators we should trim them off first.

**Comment by ouit0408** [comment]
implemented review comments and updated PR

File: portal/portal-api/api/src/java/org/sakaiproject/portal/api/Redirect.java [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+	public String getUrl() {
+		return url;
+	}
+
+	public void setUrl(String url) {
+		this.url = url;
+	}
+
+	public boolean isAppendPath() {
```
Comment (buckett) [inline portal/portal-api/api/src/java/org/sakaiproject/portal/api/Redirect.java:19]: Really good point to explain what append path means here.

File: portal/portal-api/api/src/java/org/sakaiproject/portal/api/Redirect.java [hunk 4]
```diff
@@ -0,0 +1,26 @@
+package org.sakaiproject.portal.api;
+
+/**
+ * This is to store redirect url and appendPath information
+ * for redirect of a site in hierarchy and it's subsites
+ */
+public class Redirect {
+	private String url;
```
Comment (buckett) [inline portal/portal-api/api/src/java/org/sakaiproject/portal/api/Redirect.java:8]: Something like this is an ideal case to be imutable. We know everything about  it when we're creating it and we should never need to change it later on. Having a constructor that takes all the arguments and final member variables is a good way todo this.

File: tetraelf-hierarchy/hierarchy-portal/impl/src/java/org/sakaiproject/portal/impl/HierarchySiteNeighbourhoodService.java [hunk 5]
```diff
@@ -191,17 +193,51 @@ public String parseSiteAlias(String alias) {
...
+					int index = siteId.lastIndexOf(PortalHierarchyService.SEPARATOR);
+					if(index > 0){
+						String firstPart = siteId.substring(0, index);
+						String tempString = PortalHierarchyService.SEPARATOR + siteId.substring(index + 1);
+						//save the trimmed path if there is a subsite redirect
+						trimmedPath = (StringUtils.isBlank(trimmedPath)) ? tempString : (tempString + trimmedPath) ;
+						siteId = firstPart;
+					}
+				}
+				else{
```
Comment (buckett) [inline tetraelf-hierarchy/hierarchy-portal/impl/src/java/org/sakaiproject/portal/impl/HierarchySiteNeighbourhoodService.java:223]: Formatting doesn't match the surrounding lines.

File: tetraelf-hierarchy/hierarchy-portal/impl/src/java/org/sakaiproject/portal/impl/HierarchySiteNeighbourhoodService.java [hunk 6]
```diff
@@ -191,17 +193,51 @@ public String parseSiteAlias(String alias) {
...
+				//if node null, check for the redirect of subsites
+				if(node == null){
+					//get the index of last occurrence of Separator in the path
+					int index = siteId.lastIndexOf(PortalHierarchyService.SEPARATOR);
+					if(index > 0){
+						String firstPart = siteId.substring(0, index);
+						String tempString = PortalHierarchyService.SEPARATOR + siteId.substring(index + 1);
+						//save the trimmed path if there is a subsite redirect
+						trimmedPath = (StringUtils.isBlank(trimmedPath)) ? tempString : (tempString + trimmedPath) ;
+						siteId = firstPart;
```
Comment (buckett) [inline tetraelf-hierarchy/hierarchy-portal/impl/src/java/org/sakaiproject/portal/impl/HierarchySiteNeighbourhoodService.java:220]: Generally bad style to re-assign parameters to methods. Better to create a new variable and set it to the parameter at the start of the loop.

**Comment by buckett** [comment]
None of my comments are showstoppers, they would be good to tidy up, but merging so that we get it into the build.

**Pull request merged**